### PR TITLE
Fix beta for negative args

### LINKED ISSUE #149: beta(-n,n) gives NaN, when n is an integer
Beta function at negative integer argument x=n should give ((-1)^n)/n, but instead it gives NaN.The evaluation at negative integers must be done by taking limits close to -n, where the pole at negative integers cancels out

### PR DESCRIPTION
**Pull request by Sumeghtech**
Based on SciPy implementation : https://github.com/scipy/scipy/blob/0cff7a5fe6226668729fc2551105692ce114c2b3/scipy/special/cephes/beta.c#L61

Fix #149 
cc @simonbyrne

### CODE REVIEW

File: src/gamma.jl [hunk 1]
```diff
@@ -720,6 +742,31 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
 function beta end
 
 function beta(x::Real, w::Real)
```
Comment (simonbyrne) [inline src/gamma.jl]: lets change the arguments to `a,b` to be consistent with the incomplete beta ones.

File: src/gamma.jl [hunk 2]
```diff
@@ -712,6 +712,28 @@ loggamma(z::Complex{T}) where {T<:Union{Float32,Float16}} = Complex{T}(loggamma(
 
 gamma(z::Complex) = exp(loggamma(z))
 
+#Special case for negative integer argument
+function beta_neg(a::Integer, b::Float64)
+    if b == trunc(Int,b) && 1-a-b > 0
+        sgn = ((trunc(Int,b))%2 == 0) ? 1 : -1
+        return sgn* beta(1-a-b,b)
+    else
+        return error("overflow")
```
Comment (simonbyrne) [inline src/gamma.jl]: just return a `NaN` for now.

File: src/gamma.jl [hunk 3]
```diff
@@ -720,6 +742,31 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
 function beta end
 
 function beta(x::Real, w::Real)
+    if x <= 0.0
+        if x == floor(x)
+            if x == trunc(Int,x)
+                return beta_neg(trunc(Int,x), w)
```
Comment (simonbyrne) [inline src/gamma.jl]: once you do the simplifications above, it might be easier to just inline this.

File: src/gamma.jl [hunk 4]
```diff
@@ -712,6 +712,28 @@ loggamma(z::Complex{T}) where {T<:Union{Float32,Float16}} = Complex{T}(loggamma(
...
+        sgn = ((trunc(Int,b))%2 == 0) ? 1 : -1
+        return sgn* beta(1-a-b,b)
+    else
+        return error("overflow")
+    end
+end
+
+#asymptotic expansion for log(B(a,b)) for |a| >> |b|
+function beta_asymp(a::Float64, b::Float64)
+    r, sgn = loggamma(b)
```
Comment (simonbyrne) [inline src/gamma.jl]: did you mean `logabsgamma` here?

File: test/runtests.jl [hunk 5]
```diff
@@ -802,6 +802,7 @@ end
     @test beta(3,5) ≈ 1/105
     @test logbeta(5,4) ≈ log(beta(5,4))
     @test beta(5,4) ≈ beta(4,5)
+    @test beta(-2.0,2.0) ≈ 0.5
```
Comment (simonbyrne) [inline test/runtests.jl]: add a test to check the asymptotic expansion.

**Comment by Sumeghtech** [comment]
Done @simonbyrne

**Comment by Sumeghtech** [comment]
Any comments @simonbyrne ?

File: src/gamma.jl [hunk 6]
```diff
@@ -719,18 +719,61 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
...
 
-function beta(x::Real, w::Real)
-    yx, sx = logabsgamma(x)
-    yw, sw = logabsgamma(w)
-    yxw, sxw = logabsgamma(x+w)
-    return exp(yx + yw - yxw) * (sx*sw*sxw)
+function beta(a::Real, b::Real)
+    #Special case for negative integer argument
+    if a <= 0.0
+        if a == floor(a)
```
Comment (simonbyrne) [inline src/gamma.jl]: isn't this equivalent to `isinteger(a)`?

File: src/gamma.jl [hunk 7]
```diff
@@ -719,18 +719,61 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
...
+            if isinteger(a)
```
Comment (simonbyrne) [inline src/gamma.jl]: Can combine some of the `if` statements
```suggestion
            if isinteger(a) && isinteger(b) && 1-a-b > 0
```

File: src/gamma.jl [hunk 8]
```diff
@@ -719,18 +719,61 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
...
+                if isinteger(b) && 1-a-b > 0
+                    sgn = (isinteger(b/2)) ? 1 : -1
```
Comment (simonbyrne) [inline src/gamma.jl]:
```suggestion
                    sgn = isinteger(b/2) ? 1 : -1
```
remove unneeded set of parentheses

File: src/gamma.jl [hunk 9]
```diff
@@ -719,18 +719,61 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
...
+                    return sgn* beta(1-a-b,b)
+                else
+                    return NaN
+                end
+            else
+                return NaN
+            end
+        end
+    end
+    if abs(a) < abs(b)
```
Comment (simonbyrne) [inline src/gamma.jl]: if we've reached here, they should both be positive

File: src/gamma.jl [hunk 10]
```diff
@@ -719,18 +719,61 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
...
+        a,b = b,a
+    end
+    #asymptotic expansion for log(B(a,b)) for |a| >> |b|
+    if abs(a) > 1e5*abs(b) && abs(a) > 1e5
```
Comment (simonbyrne) [inline src/gamma.jl]: could we use `loggammadiv` here? i.e. 
```julia
if a > 8
    return exp(loggammadiv(b,a) + loggamma(b))
end
```
?

**Comment by Sumeghtech** [comment]
Is it good to go now?
cc @simonbyrne

File: src/gamma.jl [hunk 11]
```diff
@@ -719,18 +719,48 @@ Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)
...
+            sgn = isinteger(a/2) ? 1 : -1
+            return sgn* beta(1-a-b,a)
+        end
+    end
+    if a < b
+        a,b = b,a
+    end
+    #asymptotic expansion for log(B(a,b)) for |a| >> |b|
+    if abs(a) > 1e5*abs(b) && abs(a) > 1e5
+        if a > 8
```
Comment (simonbyrne) [inline src/gamma.jl]: won't this condition always be true?

**Comment by simonbyrne** [comment]
Should be able to fix `logbeta` as well.

File: src/gamma.jl (hunk 11 above)
Comment (Sumeghtech) [reply src/gamma.jl]: Oh sorry yes! My mistake,

**Comment by Sumeghtech** [comment]
Done @simonbyrne !

**Comment by Sumeghtech** [comment]
ping @simonbyrne

**Pull request merged**